src/NServiceBus.Core/DataBus/FileShareDataBusImplementation.cs

```diff
@@ -53,15 +55,10 @@ public string Put(Stream stream, TimeSpan timeToBeReceived)
 
 			Directory.CreateDirectory(Path.GetDirectoryName(filePath));
 
-			using (var output = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write, FileShare.Read))
+			using (var output = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write, FileShare.Read, 4096, FileOptions.Asynchronous))
 			{
-				var buffer = new byte[32 * 1024];
-				int read;
-
-				while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
-				{
-					output.Write(buffer, 0, read);
-				}
+			    const int bufferSize = 32 * 1024;
+                await stream.CopyToAsync(output, bufferSize);
```

Comment:
Do we need a [`FlushAsync`](https://msdn.microsoft.com/en-us/library/hh193384(v=vs.110).aspx) here as well?